File: server/src/routes/playlist.ts
import { Router, Request, Response } from 'express';
import { addItemToPlaylist, queryPlaylistItems, deletePlaylistItem, clearPlaylist, updatePlaylistItem, updatePlayStatus } from '../db/queries/playlist';
import { getRoomPlayStatus, updateRoomPlayStatus, createRoomPlayStatus } from '../db/queries/roomPlayStatus';
import { PlayStatus } from '../models/PlaylistItem';
import logger from '../config/logger';

const router = Router();

router.post('/add', async (req: Request, res: Response) => {
  const { title, urls } = req.body;

  const cookiesJson = JSON.parse(req.cookies.userInfo);
  const roomId = cookiesJson.roomId;
  // validate roomId, title, urls
  if (!roomId || !title || !urls) {
    res.status(400).json({ error: 'Invalid request body' });
    return;
  }
  try {
    const playlistItemId = await addItemToPlaylist(roomId, title, urls);
    res.json({ 
      message: 'Item added to playlist',
      playlistItemId
    });
  } catch (error) {
    res.status(500).json({ message: 'Internal server error' });
  }
});

router.get('/query', async (req: Request, res: Response) => {
  // const roomId = parseInt(req.query.roomId as string);
  const cookiesJson = JSON.parse(req.cookies.userInfo);
  const roomId = cookiesJson.roomId;
  if (isNaN(roomId)) {
    res.status(400).json({ error: 'Invalid roomId' });
    return;
  }

  let playlistItemId: number | undefined;
  let playStatus: PlayStatus | undefined;
  if (req.query.playlistItemId) {
    playlistItemId = parseInt(req.query.playlistItemId as string);
    if (isNaN(playlistItemId)) {
      res.status(400).json({ error: 'Invalid playlistItemId' });
      return;
    }
  }
  if (req.query.playStatus) {
    playStatus = req.query.playStatus as PlayStatus;
    if (!Object.values(PlayStatus).includes(playStatus)) {
      res.status(400).json({ error: 'Invalid playStatus' });
      return;
    }
  }

  try {
    // if didn't specify  playStatus, query PLAYING items and NEW items
    if (!playStatus) {
      const playingItems = await queryPlaylistItems(roomId, undefined, PlayStatus.PLAYING);
      const newItems = await queryPlaylistItems(roomId, undefined, PlayStatus.NEW);
      const items = playingItems.concat(newItems);
      res.json(items);
    }
    else {
      const items = await queryPlaylistItems(roomId, playlistItemId, playStatus);
      res.json(items);
    }
  } catch (error) {
    res.status(500).json({ message: 'Internal server error' });
  }
});

router.delete('/delete', async (req: Request, res: Response) => {
  const { playlistItemId } = req.body;
  if (!playlistItemId) {
    res.status(400).json({ error: 'Invalid request body' });
    return;
  }

  try {
    // delete playlist item and its video sources
    await deletePlaylistItem(playlistItemId);
    res.json({ message: 'Item deleted from playlist' });
  } catch (error) {
    res.status(500).json({ message: 'Internal server error' });
  }
});

router.delete('/clear', async (req: Request, res: Response) => {
  // const { roomId } = req.body;
  const cookiesJson = JSON.parse(req.cookies.userInfo);
  const roomId = cookiesJson.roomId;
  if (isNaN(roomId)) {
    res.status(400).json({ error: 'Invalid roomId' });
    return;
  }

  try {
    // clear playlist items and their video sources
    await clearPlaylist(roomId);
    res.json({ message: 'Playlist cleared' });
  } catch (error) {
    logger.error('Failed to clear playlist:', error);
    res.status(500).json({ message: 'Internal server error' });
  }
});

router.post('/updateOrder', async (req: Request, res: Response) => {
  const { orderIndexList } = req.body;  // array of { playlistItemId: number, orderIndex: number }
  if (!Array.isArray(orderIndexList)) {
    res.status(400).json({ error: 'Invalid request body' });
    return;
  }
  try {
    orderIndexList.forEach((item: any) => {
      if (typeof item.playlistItemId !== 'number' || typeof item.orderIndex !== 'number') {
        res.status(400).json({ error: 'Invalid request body' });
        return;
      }
      updatePlaylistItem(item.playlistItemId, undefined, undefined, item.orderIndex); // update orderIndex only
    });
    res.json({ message: 'Order updated' });
  }
  catch (error) {
    logger.error('Failed to update order:', error);
    res.status(500).json({ message: 'Internal server error' });
  }
});

router.post('/switch', async (req: Request, res: Response) => {
  const cookiesJson = JSON.parse(req.cookies.userInfo);
  const roomId = cookiesJson.roomId;
  const { playlistItemId } = req.body;
  if (!playlistItemId) {
    res.status(400).json({ error: 'Invalid request body' });
    return;
  }

  try { 
    // set all playing items to finished and the new item to playing
    const playingItems = await queryPlaylistItems(roomId, undefined, PlayStatus.PLAYING);
    playingItems.forEach(async (item) => {
      await updatePlayStatus(item.id, PlayStatus.FINISHED);
    });
    await updatePlayStatus(playlistItemId, PlayStatus.PLAYING);

    // update room play status
    const playStatus = await getRoomPlayStatus(roomId);
    if (playStatus) {
      await updateRoomPlayStatus(roomId, { paused: false, time: 0, timestamp: Date.now(), videoId: playlistItemId });
    }
    else {
      await createRoomPlayStatus(roomId, false, 0, Date.now(), playlistItemId);
    }
    // TODO: broadcast the play status to all clients
    res.json({ message: 'Playlist item switched' });
  } catch (error) {
    logger.error('Failed to switch playlist item:', error);
    res.status(500).json({ message: 'Internal server error' });
  }
});

export default router;

File: server/internal/handlers/playlist.go
package handlers

import (
	"fmt"
	"net/http"
	"sync-player-server/internal/config"
	"sync-player-server/internal/database"
	"sync-player-server/internal/middleware"
	"sync-player-server/internal/models"
	"sync-player-server/internal/sync"
	"time"

	"github.com/gin-gonic/gin"
)

// PlaylistAdd adds an item to the playlist
func PlaylistAdd(c *gin.Context) {
	var req struct {
		Title   string                        `json:"title" binding:"required"`
		Sources []database.VideoSourceInput `json:"sources" binding:"required,min=1,dive"`
	}

	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid request body"})
		return
	}

	userInfo, ok := middleware.GetUserInfo(c)
	if !ok {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "Authentication required"})
		return
	}

	playlistItemID, err := database.AddItemToPlaylist(userInfo.RoomID, req.Title, req.Sources)
	if err != nil {
		config.Logger.Errorf("Failed to add playlist item: %v", err)
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Internal server error"})
		return
	}

	syncManager := sync.GetSyncManager()
	if syncManager != nil {
		syncManager.Broadcast(userInfo.RoomID, sync.SyncMessage{
			Type: "updatePlaylist",
		}, []uint{userInfo.UserID})
	}

	c.JSON(http.StatusOK, gin.H{
		"message":        "Item added to playlist",
		"playlistItemId": playlistItemID,
	})
}

// PlaylistQuery queries playlist items
func PlaylistQuery(c *gin.Context) {
	userInfo, ok := middleware.GetUserInfo(c)
	if !ok {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "Authentication required"})
		return
	}

	var playlistItemID *uint
	var playStatus *models.PlayStatus

	if idStr := c.Query("playlistItemId"); idStr != "" {
		var id uint
		if _, err := fmt.Sscanf(idStr, "%d", &id); err == nil {
			playlistItemID = &id
		}
	}

	if statusStr := c.Query("playStatus"); statusStr != "" {
		status := models.PlayStatus(statusStr)
		playStatus = &status
	}

	items, err := database.QueryPlaylistItems(userInfo.RoomID, playlistItemID, playStatus)
	if err != nil {
		config.Logger.Errorf("Failed to query playlist items: %v", err)
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Internal server error"})
		return
	}

	// If no playStatus filter is specified, filter out finished items on the server side
	if playStatus == nil {
		filteredItems := make([]models.PlaylistItem, 0)
		for _, item := range items {
			if item.PlayStatus != models.PlayStatusFinished {
				filteredItems = append(filteredItems, item)
			}
		}
		c.JSON(http.StatusOK, filteredItems)
		return
	}

	c.JSON(http.StatusOK, items)
}

// PlaylistDelete deletes a playlist item
func PlaylistDelete(c *gin.Context) {
	var req struct {
		PlaylistItemID uint `json:"playlistItemId" binding:"required"`
	}

	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid request body"})
		return
	}

	userInfo, ok := middleware.GetUserInfo(c)
	if !ok {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "Authentication required"})
		return
	}

	if err := database.DeletePlaylistItem(req.PlaylistItemID); err != nil {
		config.Logger.Errorf("Failed to delete playlist item: %v", err)
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Internal server error"})
		return
	}

	syncManager := sync.GetSyncManager()
	if syncManager != nil {
		syncManager.Broadcast(userInfo.RoomID, sync.SyncMessage{
			Type: "updatePlaylist",
		}, []uint{userInfo.UserID})
	}

	c.JSON(http.StatusOK, gin.H{"message": "Item deleted from playlist"})
}

// PlaylistClear clears all playlist items
func PlaylistClear(c *gin.Context) {
	userInfo, ok := middleware.GetUserInfo(c)
	if !ok {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "Authentication required"})
		return
	}

	if err := database.ClearPlaylist(userInfo.RoomID); err != nil {
		config.Logger.Errorf("Failed to clear playlist: %v", err)
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Internal server error"})
		return
	}

	syncManager := sync.GetSyncManager()
	if syncManager != nil {
		syncManager.Broadcast(userInfo.RoomID, sync.SyncMessage{
			Type: "updatePlaylist",
		}, []uint{userInfo.UserID})
	}

	c.JSON(http.StatusOK, gin.H{"message": "Playlist cleared"})
}

// PlaylistUpdateOrder updates playlist order
func PlaylistUpdateOrder(c *gin.Context) {
	var req struct {
		OrderIndexList []database.OrderIndexUpdate `json:"orderIndexList" binding:"required,min=1,dive"`
	}

	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid request body"})
		return
	}

	userInfo, ok := middleware.GetUserInfo(c)
	if !ok {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "Authentication required"})
		return
	}

	if err := database.UpdatePlaylistOrderBatch(req.OrderIndexList); err != nil {
		config.Logger.Errorf("Failed to update playlist order: %v", err)
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Internal server error"})
		return
	}

	syncManager := sync.GetSyncManager()
	if syncManager != nil {
		syncManager.Broadcast(userInfo.RoomID, sync.SyncMessage{
			Type: "updatePlaylist",
		}, []uint{userInfo.UserID})
	}

	c.JSON(http.StatusOK, gin.H{"message": "Order updated"})
}

// PlaylistSwitch switches to a playlist item
func PlaylistSwitch(c *gin.Context) {
	var req struct {
		PlaylistItemID uint `json:"playlistItemId" binding:"required"`
	}

	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid request body"})
		return
	}

	userInfo, ok := middleware.GetUserInfo(c)
	if !ok {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "Authentication required"})
		return
	}

	// Mark all currently playing items as finished
	playingStatus := models.PlayStatusPlaying
	playingItems, err := database.QueryPlaylistItems(userInfo.RoomID, nil, &playingStatus)
	if err != nil {
		config.Logger.Errorf("Failed to query playing items: %v", err)
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Internal server error"})
		return
	}

	for _, item := range playingItems {
		if err := database.UpdatePlayStatus(item.ID, models.PlayStatusFinished); err != nil {
			config.Logger.Errorf("Failed to update play status: %v", err)
		}
	}

	// Set the requested item to playing
	if err := database.UpdatePlayStatus(req.PlaylistItemID, models.PlayStatusPlaying); err != nil {
		config.Logger.Errorf("Failed to update play status: %v", err)
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Internal server error"})
		return
	}

	// Update or create room play status
	_, err = database.GetRoomPlayStatus(userInfo.RoomID)
	if err != nil {
		database.CreateRoomPlayStatus(userInfo.RoomID, false, 0, time.Now().UnixMilli(), req.PlaylistItemID)
	} else {
		database.UpdateRoomPlayStatus(userInfo.RoomID, map[string]any{
			"paused":    false,
			"time":      0.0,
			"timestamp": time.Now().UnixMilli(),
			"video_id":  req.PlaylistItemID,
		})
	}

	// Always broadcast playlist update to sync all clients
	syncManager := sync.GetSyncManager()
	if syncManager != nil {
		syncManager.Broadcast(userInfo.RoomID, sync.SyncMessage{
			Type: "updatePlaylist",
		}, []uint{userInfo.UserID})
	}

	c.JSON(http.StatusOK, gin.H{"message": "Playlist item switched"})
}
